clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -467,148 +483,178 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
-        }
-      }
-    }
+// This type manages the list of device functions and recursive functions, as
+// well as an entry point for attribute collection, for the translation unit
+// during MarkDevices. On construction this type makes sure that all of the
+// root-device functions(that is, those marked with SYCL_EXTERNAL) are
+// collected.  On destruction it manages and runs the diagnostics required. When
+// processing individual kernel/external functions, the
+// SingleDeviceFunctionTracker type updates this type.
+class DeviceFunctionTracker {
+  friend class SingleDeviceFunctionTracker;
+  CallGraph CG;
+  Sema &SemaRef;
+  // The list of functions used on the device, kept so we can diagnose on them
+  // later.
+  llvm::SmallPtrSet<FunctionDecl *, 16> DeviceFunctions;
+  llvm::SmallPtrSet<FunctionDecl *, 16> RecursiveFunctions;
+
+  void CollectSyclExternalFuncs() {
+    for (CallGraphNode::CallRecord Record : CG.getRoot()->callees())
+      if (auto *FD = dyn_cast<FunctionDecl>(Record.Callee->getDecl()))
+        if (FD->hasBody() && FD->hasAttr<SYCLDeviceAttr>())
+          SemaRef.addSyclDeviceDecl(FD);
   }
 
-  // Traverses over CallGraph to collect list of attributes applied to
-  // functions called by SYCLKernel (either directly and indirectly) which needs
-  // to be propagated down to callers and applied to SYCL kernels.
-  // For example, reqd_work_group_size, vec_len_hint, reqd_sub_group_size
-  // Attributes applied to SYCLKernel are also included
-  // Returns the kernel body function found during traversal.
-  FunctionDecl *
-  CollectPossibleKernelAttributes(FunctionDecl *SYCLKernel,
-                                  llvm::SmallVector<Attr *, 4> &Attrs) {
-    typedef std::pair<FunctionDecl *, FunctionDecl *> ChildParentPair;
-    llvm::SmallPtrSet<FunctionDecl *, 16> Visited;
-    llvm::SmallVector<ChildParentPair, 16> WorkList;
-    WorkList.push_back({SYCLKernel, nullptr});
-    FunctionDecl *KernelBody = nullptr;
+  CallGraphNode *getNodeForKernel(FunctionDecl *Kernel) {
+    assert(CG.getNode(Kernel) && "No call graph entry for a kernel?");
+    return CG.getNode(Kernel);
+  }
 
-    while (!WorkList.empty()) {
-      FunctionDecl *FD = WorkList.back().first;
-      FunctionDecl *ParentFD = WorkList.back().second;
-
-      // To implement rounding-up of a parallel-for range the
-      // SYCL header implementation modifies the kernel call like this:
-      // auto Wrapper = [=](TransformedArgType Arg) {
-      //  if (Arg[0] >= NumWorkItems[0])
-      //    return;
-      //  Arg.set_allowed_range(NumWorkItems);
-      //  KernelFunc(Arg);
-      // };
-      //
-      // This transformation leads to a condition where a kernel body
-      // function becomes callable from a new kernel body function.
-      // Hence this test.
-      if ((ParentFD == KernelBody) && isSYCLKernelBodyFunction(FD))
-        KernelBody = FD;
+  void
+  AddSingleFunctions(const llvm::SmallPtrSetImpl<FunctionDecl *> &DevFuncs,
```

Comment:
Add single functions (plural)? Maybe it's just been a long day, but that sounds weird to me when we hand this two containers of functions.

Reply:
No, you're right... I meant for 'functions' to refer to the parameters, but it sounds stupid that way.